system/lib/wasmfs/thread_utils.h

```diff
@@ -66,93 +66,158 @@ class SyncToAsync {
+      // invokers waiting to give us work, so `notify_all` to make sure our
+      // invoker wakes up.
+      parent->workCount++;
+      assert(parent->workCount && "workCount overflowed!");
+      parent->condition.notify_all();
+
       // Look for more work. Doing this asynchronously ensures that we continue
       // after the current call stack unwinds (avoiding constantly adding to the
       // stack, and also running any remaining code the caller had, like
       // destructors). TODO: add an option to do a synchronous call here in some
       // cases, which would avoid the time delay caused by a browser setTimeout.
       emscripten_async_call(threadIter, arg, 0);
-    });
+    };
+
     // Run the work function the user gave us. Give it a pointer to the resume
-    // function.
-    work(parent->resume.get());
+    // function, which it will be responsible for calling when it's done.
+    work(&parent->resume);
   }
 
 public:
-  SyncToAsync() : childLock(mutex) {
-    // The child lock is associated with the mutex, which takes the lock as we
-    // connect them, and so we must free it here so that the child can use it.
-    // Only the child will lock/unlock it from now on.
-    childLock.unlock();
-
-    // Create the thread after the lock is ready.
-    thread = std::make_unique<std::thread>(threadMain, this);
+  SyncToAsync() : workerLock(mutex, std::defer_lock) {
+    // Acquire the lock before spawning the worker thread so that we can wait
+    // for it to notify us that it has succefully started up. If we didn't lock
+    // before spawning it, we could miss that notification. If we didn't wait
+    // for it to start up at all, we could accidentally send it work before it
+    // is ready.
+    std::unique_lock<std::mutex> lock(mutex);
+    thread = std::thread(threadMain, this);
+    condition.wait(lock, [&]() { return state == Waiting; });
   }
 
   ~SyncToAsync() {
-    // Wake up the child to tell it to quit.
-    invoke([&](Callback func){
-      quit = true;
-      (*func)();
-    });
-
-    thread->join();
+    std::unique_lock<std::mutex> lock(mutex);
+
+    // We are destructing the SyncToAsync object, so we should not be racing
+    // with other threads trying to perform more `invoke`s. There should
+    // therefore not be any work available.
+    assert(state == Waiting);
+
+    // Wake the worker and tell it to quit. Be ready to join it when it does.
+    // There shouldn't be other invokers waiting to send work since we are
+    // destructing the SyncToAsync, so just use `notify_one`.
+    state = ShouldExit;
+    condition.notify_one();
+
+    // Unlock to allow the worker to wake up and exit.
+    lock.unlock();
+    thread.join();
   }
 };
 
 void SyncToAsync::invoke(std::function<void(Callback)> newWork) {
-  // Use the invokeMutex to prevent more than one invoke being in flight at a
-  // time, so that this is usable from multiple threads safely.
-  std::lock_guard<std::mutex> invokeLock(invokeMutex);
```

Comment:
I'm not sure I understand your design yet, as wasn't it simpler to have a separate mutex for invokers? That avoids some of the corner cases with a single mutex that can be grabbed by either another invoker or the worker thread, that it looks like your new code handles using `workCount` etc.

That is, with the invokerMutex we always have exactly a single invoker "in action" at a time. Other attempted invokes simply wait for when they can grab invokerMutex. Then the other mutex is purely for communication between the single active invoker and the worker.

Reply:
Good point that the separate invoker mutex avoided the need for `workCount`. On the other hand, I think that once we have actual async backends we will want to allow multiple async work items to be in flight at once (e.g. to issue separate async reads to separate files simultaneously), which will require multiple async invokers to be waiting at once. I think there is probably some value in having a design now that is similar to what we will need in the future.

Reply:
Interesting thought. What kind of idea did you have in mind for async work items here?

(For wasmfs I hadn't even thought of that as read/write are sync. But I guess for non-wasmfs stuff it could make sense?)

Reply:
For example IDB reads and writes are async and OPFS opens are async (although I don't know the particular details there). If two threads are each trying to open a new OPFS backed file, they shouldn't have to wait for each other even if they proxy those opens to the same background worker thread.